### PR TITLE
Converted to Portable library

### PR DESCRIPTION
**Pull request by Inrego**
I've converted the project to a portable library, in order to be able to target both UWP 10 and .NET 4.5

I've tested that the compiled .dll can be referenced in both a .NET 4.5 project and a UWP 10 project. I have also tested that the resulting .dll works in .NET 4.5 by successfully writing a LED Matrix to the Nuimo device. I have not yet tested the .dll in any UWP projects, but as Portable libraries are exactly for this kind of stuff, I can't imagine why it shouldn't work. Still think it should be tested before going live with the change, though.

### CODE REVIEW

**Comment by larsblumberg** [comment]
Awesome – thanks! We'll have a look into it

**Comment by Inrego** [comment]
For those not familiar with Portable Libraries (since they are a rather new concept), I'll give a brief explanation:

You are able to select which frameworks you want to target.

Change Targets

The available api's are then the intersection of the selected frameworks. As you can see in above screenshot, it automatically changes the targets by adding targets that support the same set of api's as the ones I currently selected.

In fact, I had to make a minor change to `NuimoBluetoothController.cs` since `Encoding.ASCII` was not available in all the targeted frameworks. Also, the only overload of `GetString` that was available across all the selected targets, was the one where you provide index and length of the bytes you want to read.

That means the class library now actually supports the following targets:

Available Targets

I do have my doubts about the bluetooth api's on Xamarin, it will probably just act as if there is no bluetooth controller. But .NET Framework 4.5 does work and that's what was important to me. Actually, it's possible to target .NET Framework 4.0 by removing ASP.NET Core 1.0, but I'll leave it up to you guys to do the nitpicking about which frameworks to target :smiley: 

I learned about this from a free course on Xamarin University. While we are not working with Xamarin/Android/iOS in this case, the same methodology for targeting multiple platforms are the same. I can recommend the course. It can be found here: [Introduction to Cross-Platform Mobile Development (XAM110)](https://elearning.xamarin.com/forms/xam110/) (you probably need to register and sign in before you can use that link)

**Comment by hansmbakker** [comment]
@Inrego on Xamarin (Android / iOS, but UWP not yet) you can maybe use https://github.com/xabre/xamarin-bluetooth-le/

**Comment by Inrego** [comment]
It's not that I wanted to develop for Xamarin. I was just providing examples :)

File: NuimoSDK/NuimoSDK.csproj [hunk 1]
```diff
@@ -2,123 +2,54 @@
...
-  <PropertyGroup Condition=" '$(VisualStudioVersion)' == '' or '$(VisualStudioVersion)' &lt; '14.0' ">
-    <VisualStudioVersion>14.0</VisualStudioVersion>
-  </PropertyGroup>
-  <Import Project="$(MSBuildExtensionsPath)\Microsoft\WindowsXaml\v$(VisualStudioVersion)\Microsoft.Windows.UI.Xaml.CSharp.targets" />
+  <ItemGroup>
+    <Reference Include="System.Runtime.WindowsRuntime">
+      <HintPath>..\..\..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETCore\v4.5.1\System.Runtime.WindowsRuntime.dll</HintPath>
+    </Reference>
+    <Reference Include="Windows">
+      <HintPath>..\..\..\..\..\..\Program Files (x86)\Windows Kits\10\UnionMetadata\Windows.winmd</HintPath>
```
Comment (larsblumberg) [inline NuimoSDK/NuimoSDK.csproj]: These paths seem to depend on your local file system – should they be removed here?

File: NuimoSDK/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 // associated with an assembly.
-[assembly: AssemblyTitle("Nuimo SDK")]
-[assembly: AssemblyDescription("Library for Windows 10 platforms to connect and communicate with Nuimo controllers made by Senic")]
+[assembly: AssemblyTitle("NuimoSDK")]
+[assembly: AssemblyDescription("")]
 [assembly: AssemblyConfiguration("")]
-[assembly: AssemblyCompany("Senic GmbH")]
-[assembly: AssemblyProduct("Nuimo SDK")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("NuimoSDK")]
```
Comment (larsblumberg) [inline NuimoSDK/Properties/AssemblyInfo.cs]: Is it necessary to rename the product? Does your change conform to a naming convention?

File: NuimoSDK/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
 
 // General Information about an assembly is controlled through the following 
 // set of attributes. Change these attribute values to modify the information
 // associated with an assembly.
-[assembly: AssemblyTitle("Nuimo SDK")]
-[assembly: AssemblyDescription("Library for Windows 10 platforms to connect and communicate with Nuimo controllers made by Senic")]
+[assembly: AssemblyTitle("NuimoSDK")]
+[assembly: AssemblyDescription("")]
 [assembly: AssemblyConfiguration("")]
-[assembly: AssemblyCompany("Senic GmbH")]
```
Comment (larsblumberg) [inline NuimoSDK/Properties/AssemblyInfo.cs:11]: Why did you remove the company?

File: NuimoSDK/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -1,18 +1,20 @@
...
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 // General Information about an assembly is controlled through the following 
 // set of attributes. Change these attribute values to modify the information
 // associated with an assembly.
-[assembly: AssemblyTitle("Nuimo SDK")]
-[assembly: AssemblyDescription("Library for Windows 10 platforms to connect and communicate with Nuimo controllers made by Senic")]
+[assembly: AssemblyTitle("NuimoSDK")]
+[assembly: AssemblyDescription("")]
```
Comment (larsblumberg) [inline NuimoSDK/Properties/AssemblyInfo.cs]: Accidentally removed description?

File: NuimoSDK/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -24,5 +26,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: ComVisible(false)]
-[assembly: NeutralResourcesLanguage("")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (larsblumberg) [inline NuimoSDK/Properties/AssemblyInfo.cs]: Is it necessary to specify version? It's not synchronized with the NuGet package version.

File: NuimoSDK/project.json [hunk 6]
```diff
@@ -1,16 +0,0 @@
...
-  },
-  "runtimes": {
-    "win10-arm": {},
-    "win10-arm-aot": {},
-    "win10-x86": {},
-    "win10-x86-aot": {},
-    "win10-x64": {},
-    "win10-x64-aot": {}
-  }
-}
```
Comment (larsblumberg) [inline NuimoSDK/project.json:16]: This file is not necessary?

File: NuimoSDK/NuimoBluetoothController.cs [hunk 7]
```diff
@@ -127,7 +127,7 @@ private void BluetoothGattCallback(GattCharacteristic sender, GattValueChangedEv
         private bool ReadFirmwareVersion()
         {
             return ReadCharacteristicValue(CharacteristicsGuids.FirmwareVersionGuid, bytes =>
-                DispatchOnMainAsync(() => FirmwareVersionRead?.Invoke(Encoding.ASCII.GetString(bytes)))
+                DispatchOnMainAsync(() => FirmwareVersionRead?.Invoke(Encoding.GetEncoding("ASCII").GetString(bytes, 0, bytes.Length)))
```
Comment (larsblumberg) [inline NuimoSDK/NuimoBluetoothController.cs]: Is this change necessary to make the lib compatible with all target platforms?

File: NuimoSDK/NuimoSDK.csproj (hunk 1 above)
Comment (Inrego) [inline NuimoSDK/NuimoSDK.csproj]: They are required, and will be there on any Windows pc that has the development tools for Windows 10 installed as far as I know. They are the references to the Windows 10 SDK. Without them, the project cannot compile. I will into defining a better path for them, as I see they are dependant on the location of the project on your system, so I understand your concern.

File: NuimoSDK/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (Inrego) [inline NuimoSDK/Properties/AssemblyInfo.cs]: My mistake. I made this PR by creating a whole new project and then include the source files. I forgot to overwrite the default AssemblyInfo.cs with yours. I will get around to re-including your AssemblyInfo and commit it soon.

File: NuimoSDK/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (Inrego) [inline NuimoSDK/Properties/AssemblyInfo.cs:11]: See https://github.com/getsenic/nuimo-windows/pull/8#discussion_r75353532

File: NuimoSDK/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (Inrego) [inline NuimoSDK/Properties/AssemblyInfo.cs]: See https://github.com/getsenic/nuimo-windows/pull/8#discussion_r75353532

File: NuimoSDK/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (Inrego) [inline NuimoSDK/Properties/AssemblyInfo.cs]: See https://github.com/getsenic/nuimo-windows/pull/8#discussion_r75353532

File: NuimoSDK/project.json (hunk 6 above)
Comment (Inrego) [inline NuimoSDK/project.json:16]: No it's not necessary. It's a required file for the project type you initially created. It is not used by the project type I changed to.

File: NuimoSDK/NuimoBluetoothController.cs (hunk 7 above)
Comment (Inrego) [inline NuimoSDK/NuimoBluetoothController.cs]: Yes. I mentioned this particular example here: https://github.com/getsenic/nuimo-windows/pull/8#issuecomment-237915047

It's a rather long post, so I'll quote the relevant part here:

> In fact, I had to make a minor change to NuimoBluetoothController.cs since Encoding.ASCII was not available in all the targeted frameworks. Also, the only overload of GetString that was available across all the selected targets, was the one where you provide index and length of the bytes you want to read.

**Comment by larsblumberg** [comment]
Dear @Inrego – thanks for clarifying my questions. After you found time to rebase the branch and modify the file reference from being relative to absolute using MS placeholders (I guess), we're ready to merge your PR. Thanks again!

File: NuimoSDK/NuimoSDK.csproj (hunk 1 above)
Comment (Tapanila) [reply NuimoSDK/NuimoSDK.csproj]: The HintPath should be in this format:
`<HintPath>$(MSBuildProgramFiles32)\Reference Assemblies\Microsoft\Framework\.NETCore\v4.5.1\System.Runtime.WindowsRuntime.dll</HintPath>`

`<HintPath>$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\Windows.winmd</HintPath>`

To make sure that they work without a problem when solution is moved to different folder.

**Comment by Inrego** [comment]
Sorry for the long delay. I have rebased and made the changes you asked about in your comments.

**Comment by Inrego** [comment]
Note if you get an error while compiling that says something about not referencing portable .NET it's because you still have project.lock.json in your project folder. Delete that and then it should compile fine.

**Comment by larsblumberg** [comment]
Hi @Inrego,

thanks for continuing your work on this PR. I'd like to merge but it seems you didn't properly rebase your branch on top of the master branch as your PR shows changes it shouldn't show (e.g. changes to `NuimoBluetoothController.cs`).

To properly rebase ("sync") your branch please follow these steps: https://help.github.com/articles/syncing-a-fork/, it only takes a few steps. Please don't forget to push your changes at the end.

File: NuimoSDK/NuimoSDK.nuspec [hunk 8]
```diff
@@ -1,17 +0,0 @@
...
-    <owners>Senic GmbH</owners>
-    <licenseUrl>https://raw.githubusercontent.com/getsenic/nuimo-windows/master/LICENSE</licenseUrl>
-    <projectUrl>https://github.com/getsenic/nuimo-windows</projectUrl>
-    <requireLicenseAcceptance>false</requireLicenseAcceptance>
-    <description>Library for Windows 10 platforms to connect and communicate with Nuimo controllers made by Senic</description>
-    <releaseNotes>Release 1.1.1</releaseNotes>
-    <copyright>Copyright 2016</copyright>
-    <tags></tags>
-  </metadata>
-</package>
```
Comment (larsblumberg) [inline NuimoSDK/NuimoSDK.nuspec:17]: You deleted this file not intentionally, right?

When you push your changes, please make sure that https://github.com/getsenic/nuimo-windows/pull/8 doesn't show any unwanted changes.

**Review by larsblumberg — CHANGES_REQUESTED** [review]

**Comment by Inrego** [comment]
I guess I've made too much of a mess in this branch/pull request. I will make a new PR.
